contracts/Registry.sol

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.11;
+import "./libraries/EnsHelper.sol";
+import "./libraries/CalldataValidation.sol";
+import "../../interfaces/IAllowlist.sol";
+
+/*******************************************************
+ *                      Interfaces
+ *******************************************************/
+interface IAllowlistFactory {
+  function cloneAllowlist(string memory, address) external returns (address);
+}
+
+/*******************************************************
+ *                   Main Contract Logic
+ *******************************************************/
+contract AllowlistRegistry {
+  address public factoryAddress;
+  string[] public registeredProtocols; // Array of all protocols which have successfully completed registration
+  mapping(string => address) public allowlistAddressByOriginName; // Address of protocol specific allowlist
+
+  constructor(address _factoryAddress) {
+    factoryAddress = _factoryAddress;
+  }
+
+  /**
+   * @notice Determine protocol onwer address given an origin name
+   * @param originName is the domain name for a protocol (ie. "yearn.finance")
+   * @return ownerAddress Returns the address of the domain controller if the domain is registered on ENS
+   */
+  function protocolOwnerAddressByOriginName(string memory originName)
+    public
+    view
+    returns (address ownerAddress)
+  {
+    ownerAddress = EnsHelper.ownerAddressByName(originName);
+  }
+
+  /**
+   * @notice Begin protocol registration
+   * @param originName is the domain name for a protocol (ie. "yearn.finance")
+   * @dev Only valid protocol owners can begin registration
+   * @dev Beginning registration generates a smart contract each protocol can use
+   *      to manage their conditions and validation implementation logic
+   * @dev Only fully registered protocols appear on the registration list
+   */
+  function registerProtocol(string memory originName) public {
+    // Make sure caller is protocol owner
+    address protocolOwnerAddress = protocolOwnerAddressByOriginName(originName);
+    require(
+      protocolOwnerAddress == msg.sender,
+      "Only protocol owners can register protocols"
+    );
+
+    // Make sure protocol is not already registered
+    bool protocolIsAlreadyRegistered = allowlistAddressByOriginName[
+      originName
+    ] != address(0);
+    require(
+      protocolIsAlreadyRegistered == false,
+      "Protocol is already registered"
+    );
+
+    // Clone, register and initialize allowlist
+    address allowlistAddress = IAllowlistFactory(factoryAddress).cloneAllowlist(
+      originName,
+      protocolOwnerAddress
+    );
+    allowlistAddressByOriginName[originName] = allowlistAddress;
+
+    // Register protocol
+    registeredProtocols.push(originName);
+  }
+
+  /**
+   * @notice Return a list of fully registered protocols
+   */
+  function registeredProtocolsList() public view returns (string[] memory) {
```

Comment:
Why do we need this function if `registeredProtocols` is already public? jw

Reply:
Because public array storage slots generate functions in this format `registeredProtocols(uint256)`, rather than a list. If you make a function and explicitly tell it to return `string[] memory` however it will list everything